cmd/runner/main.go

```diff
@@ -53,16 +70,294 @@ func run(args []string) error {
+	WriteBlob(ctx context.Context, uri string, r io.Reader) error
+	Scheme() blob.Scheme
+}
+
+type handler struct {
+	blob                     Blob
+	sourcePortfolioContainer string
+	destPortfolioContainer   string
+	reportContainer          string
+}
+
+func processPortfolioReq() (*task.ProcessPortfolioRequest, error) {
+	rawAssetIDs := os.Getenv("ASSET_IDS")
+	if rawAssetIDs == "" {
+		return nil, errors.New("no ASSET_IDS given")
+	}
+
+	var assetIDs []string
+	if err := json.NewDecoder(strings.NewReader(rawAssetIDs)).Decode(&assetIDs); err != nil {
+		return nil, fmt.Errorf("failed to load asset IDs: %w", err)
+	}
+
+	return &task.ProcessPortfolioRequest{
+		AssetIDs: assetIDs,
+	}, nil
+}
+
+func (h *handler) uploadDirectory(ctx context.Context, dirPath, container string) error {
+	base := filepath.Base(dirPath)
+
+	return filepath.WalkDir(dirPath, func(path string, info fs.DirEntry, err error) error {
+		if info.IsDir() {
+			return nil
+		}
+
+		// This is a file, let's upload it to the container
+		uri := blob.Join(h.blob.Scheme(), container, base, strings.TrimPrefix(path, dirPath))
+		if err := h.uploadBlob(ctx, path, uri); err != nil {
+			return fmt.Errorf("failed to upload blob: %w", err)
+		}
+		return nil
+	})
+}
+
+func (h *handler) uploadBlob(ctx context.Context, srcPath, destURI string) error {
+	srcF, err := os.Open(srcPath)
+	if err != nil {
+		return fmt.Errorf("failed to open file for upload: %w", err)
+	}
+	defer srcF.Close() // Best-effort in case something fails
+
+	if err := h.blob.WriteBlob(ctx, destURI, srcF); err != nil {
+		return fmt.Errorf("failed to write file to blob storage: %w", err)
+	}
+
+	if err := srcF.Close(); err != nil {
+		return fmt.Errorf("failed to close source file: %w", err)
+	}
+
+	return nil
+}
+
+func (h *handler) downloadBlob(ctx context.Context, srcURI, destPath string) error {
+	// Make sure the destination exists
+	if err := os.MkdirAll(filepath.Dir(destPath), 0600); err != nil {
+		return fmt.Errorf("failed to create directory to download blob to: %w", err)
+	}
+
+	destF, err := os.Create(destPath)
+	if err != nil {
+		return fmt.Errorf("failed to create dest file: %w", err)
+	}
+	defer destF.Close() // Best-effort in case something fails
 
-	h := taskrunner.New(&local.Executor{}, logger)
+	br, err := h.blob.ReadBlob(ctx, srcURI)
+	if err != nil {
+		return fmt.Errorf("failed to read raw portfolio: %w", err)
+	}
+	defer br.Close() // Best-effort in case something fails
+
+	if _, err := io.Copy(destF, br); err != nil {
+		return fmt.Errorf("failed to load raw portfolio: %w", err)
+	}
+
+	if err := br.Close(); err != nil {
+		return fmt.Errorf("failed to close blob reader: %w", err)
+	}
+
+	if err := destF.Close(); err != nil {
+		return fmt.Errorf("failed to close dest file: %w", err)
+	}
+
+	return nil
+}
+
+func (h *handler) processPortfolio(ctx context.Context, req *task.ProcessPortfolioRequest) error {
+	// Load the portfolio from blob storage, place it in /mnt/raw_portfolios, where
```

Comment:
s/portfolio/portfolios/(many places)

Reply:
That's actually a question I had for you about how we want to architect async task running. I was originally thinking that the async runner would get a "portfolio group ID" or whatever, but that means it needs to have database access, to translate from "portfolio group X" to "list of assets in portfolios"

I'd like it to be as dumb as possible because it's a large Docker image with code from N different teams in N different languages, so ideally we could sculpt the interface to be "here are some files in blob store, do stuff to them and write them to this other blob store", which  minimizes the permissions that we need to give the runner.

My question here is what the end-user experience/data model looks like and if that is a reasonable/achievable goal. Are there reasons we _can't_ process portfolios within a portfolio group individually at this stage? Let's chat if it's not a simple answer.